### PR TITLE
Us001 view job roles

### PR DESCRIPTION
**Pull request by Fluffyguy611**
### 🛠 Changes being made

Here give examples of the changes you've made in this pull request. Include an itemized list if you can. It'll help the reviewer

### ✨ Definition of Done (all the points need to be filled out before merging)

- [ ] Meets Acceptance Criteria
- [ ] Tested
- [ ] Code Review made by team 2 members
- [ ] UX approved
- [ ] Product Owner Approved

### 🧪 Non-Functional Requirements

- [ ] Automated tests should cover acceptance criterias (Unit tests, Integration/API/Component tests)
- [ ] All input must be validated (client side & API)
- [ ] Sensitive data should not be stored in the URL
- [ ] No secrets in codesource

### CODE REVIEW

File: product-web-app/tsconfig.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
     "moduleResolution": "Node16",
     /* Modules */
-    "module": "ES2020",                                /* Specify what module code is generated. */
```
Comment (ljercha) [inline product-web-app/tsconfig.json]: Keep it as it was

File: product-backend/src/test/java/test/JobRoleServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatExceptionOfType;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+
+@ExtendWith(MockitoExtension.class)
+public class JobRoleServiceTest {
```
Comment (ljercha) [inline product-backend/src/test/java/test/JobRoleServiceTest.java]: - lack of formatting
- you are actually mocking the real service you want to test, you should mock service dependencies

File: product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 3]
```diff
@@ -2,25 +2,43 @@
...
+        this.jobRoleDao = jobRoleDao;
+    }
+
+    public List<JobRole> getAllJobRoles() throws SQLException, FailedToGetAllJobRolesException {
+
+        try {
+            List<JobRole> jobRoleList = jobRoleDao.getAllJobRoles();
+            return jobRoleList;
+        } catch (SQLException | DatabaseConnectionException e) {
+            System.err.println(e.getMessage());
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java]: use logger sl4j see example usage on main or here
https://github.com/ljercha/agile-project

File: product-web-app/tests/jobRoleService.test.ts [hunk 4]
```diff
@@ -60,5 +60,40 @@ describe('JobRole service', () => {
...
+
+      const result = await jobRolesService.getJobRoles();
+      expect(result).to.deep.equal(data);
+    });
+
+    it('should throw err when  ', async () => {
+      const mock = new MockAdapter(axios);
+      const jobRolesService = new JobRoleService();
+
+      mock.onGet('http://localhost:8080/api/job-roles').reply(500, data);
```
Comment (ljercha) [inline product-web-app/tests/jobRoleService.test.ts]:
```suggestion
      mock.onGet('/api/job-roles').reply(500, data);
```

File: product-web-app/views/job-roles.html [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            <h2>Our Job Roles</h2>
+          </div>
+        </div> 
+      </div> 
+    </header>
+
+    <main class="container mt-4">
+      <section id="roles" class="mt-4">
+
+      {% if roles and roles.length > 0%}
```
Comment (ljercha) [inline product-web-app/views/job-roles.html]:
```suggestion
      {% if roles and roles.length > 0 %}
```

File: product-web-app/views/job-roles.html [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  <body>
+    <header class="bg-dark text-light">
+      <div class="container">
+        <div class="row">
+          <div class="col-md-12">
+            <h2>Our Job Roles</h2>
+          </div>
+        </div> 
+      </div> 
+    </header>
```
Comment (ljercha) [inline product-web-app/views/job-roles.html]: parent already contain header

File: product-web-app/views/job-roles.html [hunk 7]
```diff
@@ -0,0 +1,41 @@
+{% extends "./parent.html" %}
+{% block content %}
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>JOB ROLE LIST</title>
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css">
+  </head>
```
Comment (ljercha) [inline product-web-app/views/job-roles.html]: you should get rid of this

**Review by ljercha — APPROVED** [review]

File: product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 8]
```diff
@@ -2,25 +2,46 @@
...
+    }
+
+    public List<JobRole> getAllJobRoles() throws SQLException, FailedToGetAllJobRolesException {
+
+        try {
+            List<JobRole> jobRoleList = jobRoleDao.getAllJobRoles();
+            return jobRoleList;
+        } catch (SQLException | DatabaseConnectionException e) {
+            logger.error("SQL exception! Error: {}. DatabaseConnection! Error: {}", e.getMessage(), e.getMessage());
+            throw new FailedToGetAllJobRolesException();
```
Comment (Fluffyguy611) [inline product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java]: Only one Exception will catch 
```
logger.error("Exception! Error: {}", e.getMessage());
```

File: product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java [hunk 9]
```diff
@@ -31,8 +52,7 @@ public JobRole getJobRoleById(int id) throws JobRoleDoesNotExistException, Faile
 
             return jobRoleSingleView.get();
         } catch (SQLException | DatabaseConnectionException e) {
-            System.err.println(e.getMessage());
-
+            logger.error("SQL exception! Error: {}. DatabaseConnection! Error: {}", e.getMessage(), e.getMessage());
```
Comment (Fluffyguy611) [inline product-backend/src/main/java/org/kainos/ea/api/JobRoleService.java]: Same as above

File: product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 10]
```diff
@@ -17,14 +18,27 @@
...
 
 
+    @GET
+    @Path("/job-roles")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getAllJobRoles() {
+        try {
+            return Response.ok(jobRoleService.getAllJobRoles()).build();
+        } catch (FailedToGetAllJobRolesException | SQLException e) {
+            logger.error("Failed to get Job Roles! Error: {}. SQL exception! Error: {}", e.getMessage(), e.getMessage());
```
Comment (Fluffyguy611) [inline product-backend/src/main/java/org/kainos/ea/controller/JobRoleController.java]: Same as above

File: product-backend/src/main/java/org/kainos/ea/db/JobRoleDao.java [hunk 11]
```diff
@@ -1,15 +1,37 @@
...
+    }
+
+    public List<JobRole> getAllJobRoles() throws SQLException, DatabaseConnectionException {
+        Connection c = databaseConnector.getConnection();
+        Statement st = c.createStatement();
+        ResultSet rs = st.executeQuery("SELECT JobRole.Id, JobRole.Name, JobRole.Description, JobRole.SharePointLink\n" + "FROM JobRole");
+
+        List<JobRole> jobRoleList = new ArrayList<>();
+        while (rs.next()) {
+            JobRole jobRole = new JobRole(
```
Comment (Fluffyguy611) [inline product-backend/src/main/java/org/kainos/ea/db/JobRoleDao.java:29]: Prefer Optional.of(instead of returning List<>)

**Review by BartlomiejMilewski — APPROVED** [review]

**Pull request merged**